README.pod
=head1 NAME AND VERSION

IO::BlockSync - Syncronize changed blocks

=head1 VERSION

Version 0.001

=head1 SYNOPSIS

BlockSync can some of the same stuff that bigsync (by Egor Egorov) can
- it's just written in perl.

BlockSync copies data from source file to destination file (can be a block
device) and calculates checksum on each block it copies.
On all runs after the first only the changed blocks will be copied.

Perl module:

    use IO::BlockSync;
    BlockSync( src => 'sfile', dst => 'dfile', chk => 'cfile' );

Executable:

    blocksync -s /source/file -d /destination/file

=head1 INSTALLATION

When installing this module you install the C<IO::BlockSync> Perl module
and the C<blocksync> command.

=head2 CPAN

Install directly from CPAN mirror:

    #cpan IO::BlockSync # Not available yet

Or see
L<How to install CPAN modules|http://www.cpan.org/modules/INSTALL.html>.

=head2 Local dir

You can get a copy of this module from GitHub:

    git clone https://github.com/thordreier/perl-IO-BlockSync.git

Then you can just use CPAN to install the module and take care of dependencies:

    cpan .

Or manual install (you need to fix dependencies manually):

    perl Makefile.PL
    make
    make test
    make install

=head2 Debian/Ubuntu

If you want to use as many packages from your distro as possible instead of
getting them from CPAN you can run this command before any of the ones above:

    apt-get install \
        libnamespace-clean-perl \
        liblog-log4perl-perl \
        libmoo-perl \
        libmoosex-log-log4perl-perl \
        libmoox-options-perl \
        libtry-tiny-perl

=head1 USAGE AND DOCUMENTATION

=head2 Perl module

After installing, you can find documentation for this module with the
perldoc command:

    perldoc IO::BlockSync

This documentation can also be found on
L<GitHub|https://github.com/thordreier/perl-IO-BlockSync/blob/master/pod/IO-BlockSync.pod>

=head2 Executable

After installing, you can find documentation for the executable with one
of these commands:

    blocksync --help
    blocksync --man
    perldoc $(which blocksync)

Documentation can also be found on 
L<GitHub|https://github.com/thordreier/perl-IO-BlockSync/blob/master/pod/blocksync.pod>

=head1 LICENSE AND COPYRIGHT

This software is copyright (c) 2019 by Thor Dreier-Hansen.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

Terms of the Perl programming language system itself:

=over

=item * the 
        L<GNU General Public License|http://dev.perl.org/licenses/gpl1.html>
        as published by the Free Software Foundation; either
        L<version 1|http://dev.perl.org/licenses/gpl1.html>,
        or (at your option) any later version, or

=item * the L<"Artistic License"|http://dev.perl.org/licenses/artistic.html>

=back

See L<http://dev.perl.org/licenses/> for more information.


Changes
Revision history for IO-BlockSync

0.001   2019-06-29
        First version out



Makefile.PL
use 5.010;
use strict;
use warnings FATAL => 'all';
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME => 'IO::BlockSync',
    AUTHOR =>
      q{Thor Dreier-Hansen <43479811+thordreier@users.noreply.github.com>},
    VERSION_FROM       => 'lib/IO/BlockSync.pm',
    ABSTRACT_FROM      => 'lib/IO/BlockSync.pm',
    LICENSE            => 'perl_5',
    EXE_FILES          => ['bin/blocksync'],
    MIN_PERL_VERSION   => '5.010',
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker' => '0',
    },
    BUILD_REQUIRES => {
        'Test::More' => '0',
    },
    PREREQ_PM => {
        'namespace::clean'      => '0.27',
        'Log::Log4perl'         => '1.48',
        'Moo'                   => '2.002',
        'MooseX::Log::Log4perl' => '0.47',
        'MooX::Options'         => '4.023',
        'Try::Tiny'             => '0.28',
    },
    dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => { FILES    => 'IO-BlockSync-*' },
);


misc/releasetest.sh
#!/bin/bash

set -e

function color_echo {
    echo -e "\e[33m$1\e[39m"
}

function color_cmd {
    color_echo "$1"
    sh -c "$1"
}

function copy_pod {
    cp -a $2 $2.tmp
    (
    perl -ne 'exit if $_ =~ /__END__/; print $_' < $2
    echo '__END__'
	perldoc -u $1
    ) > $2.tmp
    mv $2.tmp $2
    echo Copied POD from $1 to $2
}

function perlfiles {
    (
	find $1 -name '*.pl' -or -name '*.pm' -or -name '*.PL'
	(
	    find $1 -type f -executable | grep -v .git
	    find $1 -type f -name '*.t'
	    
	) | xargs -i egrep -l '^#!.*perl' {}
    ) | sort | uniq
}


color_echo 'Clean it all'
perl Makefile.PL
make realclean
rm -fv MANIFEST MANIFEST.bak

color_echo 'Tidying Perl Files'
for FILE in $(perlfiles .)
do
    CMD="perltidy -b -bext='/' $FILE"
    echo $CMD
    $CMD
done

color_echo 'Copy/replace POD'
copy_pod lib/IO/BlockSync/App.pm bin/blocksync

color_cmd 'perl Makefile.PL'
color_cmd 'make test'
color_cmd 'make manifest'
color_cmd 'make blibdirs'
color_cmd 'RELEASE_TESTING=1 make test'

color_echo 'Creating POD files'
PODDIR=pod
mkdir -p $PODDIR
cat > $PODDIR/README.pod <<EOF
=head1 pod directory

Files in this directory are generated by script.

POD files are generated so it is easier to read documentation directly
from GitHub website.
EOF
for PL in $(perlfiles blib)
do
    POD=$(echo $PL | sed -r 's#blib/(script|lib)/##; s#\/#-#g; s#^(.+)(\..+)$#\1#')
    POD="$PODDIR/$POD.pod"
    TMP=$(tempfile)
    if $(perldoc -u $PL > $TMP)
    then
        echo "$PL -> $POD"
        mv $TMP $POD
    else
        rm -f $TMP
    fi
done

color_echo 'Install local'
cpan .

color_echo 'Testing blocksync'
./misc/blocksynctest.sh


misc/blocksynctest.sh
#!/bin/bash

# Maybe these tests should be rewritten in Perl and placed in t/


SRC=$(tempfile)
DST=$(tempfile)
OUT=$(tempfile)
EXIT=0

function grepout {
    GOT=$(grep "<$1>" $OUT | wc -l)
    if [ $GOT -eq $2 ]
    then
	echo -e "\e[32mOK\e[39m   $1 $2/$GOT"
    else
	echo -e "\e[31mFAIL\e[39m $1 $2/$GOT"
	EXIT=1
    fi
}


function cmpsrcdst {
    blocksync -S -s $SRC -d $DST -v $6 > $OUT
    #cat $OUT
    if [ $(md5sum $SRC $DST | awk '{print $1}' | uniq | wc -l) -eq $1 ]
    then
	echo -e "\e[32mOK\e[39m   checksum"
    else
	echo -e "\e[31mFAIL\e[39m checksum"
	EXIT=1
    fi
    grepout sparse    $2
    grepout new       $3
    grepout changed   $4
    grepout unchanged $5
}


(
    dd if=/dev/zero     bs=1M   count=1 2> /dev/null
    dd if=/dev/urandom  bs=1M   count=1 2> /dev/null
) | dd of=$SRC 2> /dev/null
echo overwrite/trunc - 1 MB zero, 1 MB rand
cmpsrcdst 1 1 1 0 0


(
    dd if=/dev/zero     bs=1M   count=1 2> /dev/null
    dd if=/dev/urandom  bs=1M   count=1 2> /dev/null
    dd if=/dev/zero     bs=1M   count=1 2> /dev/null
    dd if=/dev/urandom  bs=1M   count=1 2> /dev/null
) | dd of=$SRC 2> /dev/null
echo overwrite/trunc - 1 MB zero, 1 MB rand, 1 MB zero, 1 MB rand
cmpsrcdst 1 1 1 1 1


(
    dd if=/dev/urandom  bs=1M   count=10 2> /dev/null
    dd if=/dev/zero     bs=1M   count=10 2> /dev/null
    dd if=/dev/urandom  bs=512K count=19 2> /dev/null
) | dd of=$SRC 2> /dev/null
echo overwrite/trunc - 10 MB rand, 10 MB zero, 9,5 MB rand
cmpsrcdst 1 10 16 4 0
echo nothing changed
cmpsrcdst 1 0 0 0 30


(
    dd if=/dev/zero     bs=1M   count=6 2> /dev/null
    dd if=/dev/urandom  bs=1M   count=6 2> /dev/null
) | dd of=$SRC conv=notrunc 2> /dev/null
echo overwrite/notrunc - 6 MB zero, 6 MB rand
cmpsrcdst 1 0 0 12 18


(
    dd if=/dev/zero     bs=1M   count=6  2> /dev/null
    yes | dd iflag=fullblock bs=512K count=11 2> /dev/null
) | dd of=$SRC 2> /dev/null
echo overwrite/trunc src - 6 MB zero, 5,5 MB y\\n
cmpsrcdst 2 0 0 6 6
echo nothing changed/trunc dst
cmpsrcdst 1 0 0 0 12 -t


truncate -s 6291456 $SRC
echo truncate src+dst - 6 MB
cmpsrcdst 1 0 0 0 6 -t


(
    dd if=/dev/zero     bs=1M   count=6  2> /dev/null
    yes | dd  iflag=fullblock bs=512K count=11 2> /dev/null
) | dd of=$SRC 2> /dev/null
echo overwrite/trunc src - 6 MB zero, 5,5 MB y\\n
cmpsrcdst 1 0 6 0 6

(
    dd if=/dev/zero     bs=1M   count=6  2> /dev/null
    yes | dd iflag=fullblock bs=1M   count=6 2> /dev/null
    dd if=/dev/zero     bs=1M   count=20 2> /dev/null
) | dd of=$SRC 2> /dev/null
echo overwrite/trunc src - 6 MB zero, 6 MB y\\n, 20 MB zero
cmpsrcdst 1 20 0 1 11


(
    dd if=/dev/zero     bs=512K count=1 2> /dev/null
) >> $SRC
echo append - 0,5 MB zero - not full block - will not be sparse
cmpsrcdst 1 0 1 0 32


rm -f $SRC* $DST* $OUT*

exit $EXIT


lib/IO/BlockSync/App.pm
package IO::BlockSync::App;

# Basic
use 5.010;
use strict;
use warnings FATAL => 'all';

# Build in
use English qw( -no_match_vars );
use Scalar::Util qw(blessed);

# CPAN
use Moo;
use MooX::Options;
use Log::Log4perl;
use Log::Log4perl::Level;

# This bundle
use IO::BlockSync;

################################################################

=head1 NAME

IO::BlockSync::App (Perl module) or blocksync (command)

=head1 VERSION

Version 0.001

=cut

our $VERSION = '0.001';

################################################################

=head1 SYNOPSIS

BlockSync can some of the same stuff that bigsync (by Egor Egorov) can
- it's just written in perl.

BlockSync copies data from source file to destination file (can be a block
device) and calculates checksum on each block it copies.
On all runs after the first only the changed blocks will be copied.

    blocksync -S -s /source/path -d /destination/path

=cut

################################################################

=head1 COMMAND LINE OPTIONS

=cut

################################################################

=head2 -V --version

Print version and exit

=cut

option 'version' => (
    is    => 'ro',
    short => 'V',
    doc   => 'print version and exit',
);

################################################################

=head2 -v --verbose

Print version and exit

=cut

option 'verbose' => (
    is         => 'ro',
    short      => 'v',
    repeatable => 1,
    doc        => 'be verbose',
);

################################################################

=head2 -s --src

Path to source file.

mandatory - string (containing path)

=cut

option 'src' => (
    is     => 'ro',
    format => 's',
    short  => 's',
    doc    => 'source file path',
);

################################################################

=head2 -d --dst

Destination file. If not set, then only checksum file will be updated.

optional - string (containing path)

=cut

option 'dst' => (
    is     => 'ro',
    format => 's',
    short  => 'd',
    doc    => 'destination file path',
);

################################################################

=head2 -c --chk

Path to checksum file.

mandatory - string (containing path)

=cut

option 'chk' => (
    is     => 'rwp',
    format => 's',
    short  => 'c',
    doc    => 'checksum file path',
);

################################################################

=head2 -b --bs

Block size to use in bytes.

optional - integer - defaults to 1_048_576 B (1 MB)

=cut

option 'bs' => (
    is      => 'ro',
    format  => 'i',
    short   => 'b',
    default => 1_048_576,
    doc     => 'block size',
);

################################################################

=head2 -S --sparse

Seek in dst file, instead of writing blocks only containing \0

optional - boolean - defaults to 0 (false)

=cut

option 'sparse' => (
    is      => 'ro',
    short   => 'S',
    default => 0,
    ## no critic (ValuesAndExpressions::RequireInterpolationOfMetachars)
    doc => 'seek in dst file, instead of writing blocks only containing \0',
    ## use critic
);

################################################################

=head2 -t --truncate

Truncate the destination file to same size as source file. Does not work on block devices. Will only be tried if C<data> has default value (whole file is copied).

optional - boolean - defaults to 0 (false)

=cut

option 'truncate' => (
    is      => 'ro',
    short   => 't',
    default => 0,
    doc     => 'truncate destination file',
);

################################################################

=head1 METHODS

=cut

################################################################

=head2 run

C<bin/blocksync> calls C<IO::BlockSync::App::run> to start the program.

If you just run C<blocksync> command then ignore this method.

=cut

sub run {
    my ($self) = @_;

    # Create object with data from command line and run this function again
    if ( !blessed $self) {
        return $self->new_with_options->run;
    }

    if ( $self->version ) {
        say "blocksync version $VERSION";
        exit;
    }

    if ( !$self->src ) {
        $self->options_usage( 1, '<src> must be set' );
    }

    if ( $self->dst && !$self->chk ) {
        $self->_set_chk( $self->dst . '.chk' );
    }
    elsif ( !$self->dst && !$self->chk ) {
        $self->options_usage( 1, 'at least one of <chk> or <dst> must be set' );
    }

    my $status;
    if ( $self->verbose ) {
        $status = sub {
            printf
              "Copying block=<%s> type=<%s> start=<%s> end=<%s> size=<%s>\n",
              @_, $_[3] - $_[2];
        };
        if ( $self->verbose > 1 ) {
            Log::Log4perl->init( \<<"EOF");
                log4perl.logger = DEBUG, Screen
                log4perl.appender.Screen = Log::Log4perl::Appender::Screen
                log4perl.appender.Screen.layout = Log::Log4perl::Layout::PatternLayout
                log4perl.appender.Screen.layout.ConversionPattern = %d %p %M %m%n
EOF
        }
    }
    else {
        $status = sub {
            $OUTPUT_AUTOFLUSH++;
            print( "\rCopying block " . shift );
        }
    }

    BlockSync( %{$self}, 'status' => $status, );

    say '';

    # Make Perl::Critic happy
    return;
}

################################################################

=head1 LICENSE AND COPYRIGHT

This software is copyright (c) 2019 by Thor Dreier-Hansen.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

Terms of the Perl programming language system itself:

=over

=item * the 
        L<GNU General Public License|http://dev.perl.org/licenses/gpl1.html>
        as published by the Free Software Foundation; either
        L<version 1|http://dev.perl.org/licenses/gpl1.html>,
        or (at your option) any later version, or

=item * the L<"Artistic License"|http://dev.perl.org/licenses/artistic.html>

=back

See L<http://dev.perl.org/licenses/> for more information.

=cut

1;    # End of IO::BlockSync::App


lib/IO/BlockSync.pm
package IO::BlockSync;

# Basic
use 5.010;
use strict;
use warnings FATAL => 'all';

# Build in
use Carp;
use Fcntl qw(:DEFAULT :seek);
use POSIX qw(ceil);
use Scalar::Util qw(reftype);

# CPAN
use Log::Log4perl;
use Log::Log4perl::Level;
use Moo;
use Try::Tiny;

# These two come last - in that order
use namespace::clean;
use Exporter qw(import);

################################################################

# Moo roles to implement
with('MooseX::Log::Log4perl');

# Make sure log4perl doesn't come with errors
if ( not Log::Log4perl->initialized() ) {
    Log::Log4perl->easy_init( Log::Log4perl::Level::to_priority('OFF') );
}

################################################################

=head1 NAME

IO::BlockSync - Syncronize changed blocks

=head1 VERSION

Version 0.001

=cut

our $VERSION = '0.001';

################################################################

=head1 SYNOPSIS

BlockSync can some of the same stuff that bigsync (by Egor Egorov) can
- it's just written in perl.

BlockSync copies data from source file to destination file (can be a block
device) and calculates checksum on each block it copies.
On all runs after the first only the changed blocks will be copied.

    use IO::BlockSync;

    # OOP way
    my $bs = IO::BlockSync->new(
        src => '/path/to/source/file',
        dst => '/path/to/destination/file',
        chk => '/path/to/chk/file',
    );
    $bs->run;

    # Non OOP way
    BlockSync(
        src => '/path/to/source/file',
        dst => '/path/to/destination/file',
        chk => '/path/to/chk/file',
    );

=cut

################################################################

=head1 EXPORT

=cut

our @EXPORT = qw(BlockSync);

################################################################

=head2 BlockSync

Run BlockSync non-object-oriented

=cut

sub BlockSync {
    return __PACKAGE__->new(@_)->run;
}

################################################################

=head1 ATTRIBUTES

=cut

################################################################

=head2 src

Path to source file.

mandatory - string (containing path) or filehandle

=cut

has 'src' => (
    is       => 'ro',
    required => 1,
);

################################################################

=head2 dst

Destination file. If not set, then only checksum file will be updated.

optional - string (containing path) or filehandle

=cut

has 'dst' => ( is => 'ro', );

################################################################

=head2 chk

Path to checksum file.

mandatory - string (containing path) or filehandle

=cut

has 'chk' => (
    is       => 'ro',
    required => 1,
);

################################################################

=head2 bs

Block size to use in bytes.

optional - integer - defaults to 1_048_576 B (1 MB)

=cut

has 'bs' => (
    is      => 'ro',
    default => 1_048_576,
);

################################################################

=head2 hash

Sub that retrurn hashed data.

optional - sub - defaults to sub that return MD5 hash followed by newline

=cut

has 'hash' => (
    is      => 'ro',
    default => sub {
        require Digest::MD5;
        sub {
            Digest::MD5::md5_hex(shift) . "\n";
        }
    },
);

################################################################

=head2 sparse

Seek in dst file, instead of writing blocks only containing \0

optional - boolean - defaults to 0 (false)

=cut

has 'sparse' => (
    is      => 'ro',
    default => 0,
);

################################################################

=head2 truncate

Truncate the destination file to same size as source file. Does not work on block devices. Will only be tried if C<data> has default value (whole file is copied).

optional - boolean - defaults to 0 (false)

=cut

has 'truncate' => (
    is      => 'ro',
    default => 0,
);

################################################################

=head2 data

List of areas (in bytes) inside the source file that should be looked at.
Usefull if you know excactly which blocks in src that could have changed.

data => [
    {start => 0, end => 9999},
    {start => 88888, end => 777777},
]

optional - array of hashes - defaults to "whole file"

=cut

has 'data' => (
    is      => 'ro',
    default => sub {
        [
            {
                start => 0,
                end   => 0
            }
        ]
    },
);

################################################################

=head2 status

Sub that will be run everytime a block has been read (and written).

optional - sub - default to sub doing nothing

=cut

has 'status' => (
    is      => 'ro',
    default => sub {
        sub { }
    },
);

################################################################

=head1 METHODS

=cut

################################################################

=head2 run

This is the method that starts copying data.

=cut

sub run {    ## no critic (Subroutines::ProhibitExcessComplexity)
    my $self = shift;

    my ( $srcFh, $srcClose, $dstFh, $dstClose, $chkFh, $chkClose );

    try {
        # Get file handles for source, destination and checksum files
        $srcFh = $self->_getFh( 'src', $self->src, \$srcClose, O_RDONLY );
        $chkFh =
          $self->_getFh( 'chk', $self->chk, \$chkClose, O_RDWR | O_CREAT );
        if ( $self->dst ) {
            $dstFh = $self->_getFh( 'dst', $self->dst, \$dstClose,
                O_WRONLY | O_CREAT );
        }
        else {
            $self->logger->debug('No dst file, only calculating checksums');
        }

        # Calculate hash for a block only containing ASCII 0
        my $nullHash = &{ $self->hash }( "\0" x $self->bs );

        # Get number of bytes that a hash takes up
        my $hashSize = length($nullHash);

        my $srcSeek;

        # Loop through "areas" that should be copied
        # Default i one area containing the whole source file
        foreach my $dataBlocks ( @{ $self->data } ) {

            # Start and end of this "area" (default is $start=0, $end=0)
            my $start = $dataBlocks->{start};
            my $end   = $dataBlocks->{end};

            $self->logger->debug(
                "Going to process data from <$start> to <$end>");

            # Seek to $start
            # (or the beginning of the block that $start is in,
            # if $start is not aligned with bs)
            $srcSeek = int( $start / $self->bs ) * $self->bs;
            sysseek( $srcFh, $srcSeek, SEEK_SET )
              || $self->logger->logcroak(
                "Cannot seek to block <$srcSeek> in src file");

            # Just die! Muhahaha. Or not
            my $die = 0;

            # Can be either sparse, new, unchanged or changed
            my $status;

            # Read block from source
            while ( my $srcReadSize = sysread( $srcFh, my $data, $self->bs ) ) {

                # It's ok to read a block smaller than bs if it's the last
                # block. But it's not ok if it's not the last.
                if ($die) {
                    croak 'not reading full block';
                }
                $die = $srcReadSize != $self->bs;

                # $block = block number in source with the specified block size
                my $block = $srcSeek / $self->bs;

                # We start by assuming that we should write to dst
                # - if dst is set (= we are not just calculating checksum)
                my $writeData = 1 && $dstFh;

                # We start be assuming that we should write checksum to chk
                my $writeHash = 1;

                # Calculate hash for data read from src
                my $newHash = &{ $self->hash }($data);

                # Get old hash for the same block
                sysseek( $chkFh, $block * $hashSize, SEEK_SET )
                  || $self->logger->logcroak('Cannot seek in chk file');
                my $oldHashSize = sysread( $chkFh, my $oldHash, $hashSize );

                # Test source against checksum
                if ( $oldHashSize != $hashSize || $oldHash eq "\0" x $hashSize )
                {
                    if ( $self->sparse && $newHash eq $nullHash ) {

                        # Sparse is only for new blocks
                        # Blocks that have been nulled out in source will
                        # also get nulled out in destination
                        $status    = 'sparse';
                        $writeData = 0;
                    }
                    else {
                        $status = 'new';
                    }
                }
                elsif ( $newHash eq $oldHash ) {
                    $status    = 'unchanged';
                    $writeData = 0;
                    $writeHash = 0;
                }
                else {
                    $status = 'changed';
                }

                # Write data to destination
                if ($writeData) {
                    sysseek( $dstFh, $srcSeek, SEEK_SET )
                      || $self->logger->logcroak('Cannot seek in dst file');
                    syswrite( $dstFh, $data );
                }

                # Update hash in checksum
                if ($writeHash) {
                    sysseek( $chkFh, $block * $hashSize, SEEK_SET )
                      || $self->logger->logcroak('Cannot seek in chk file');
                    syswrite( $chkFh, $newHash );
                }

                $self->logger->debug(
                    sprintf 'Block <%u> was <%s> (<%u> to <%u>)',
                    $block, $status, $srcSeek, $srcSeek + $srcReadSize - 1 );
                &{ $self->status }
                  ( $block, $status, $srcSeek, $srcSeek + $srcReadSize - 1 );

                # Next block will start here
                $srcSeek += $srcReadSize;

                # Was this the last block in this batch
                if ( $end && $srcSeek > $end ) {
                    last;
                }
            }

            # If last block is sparse, it is not enough to seek to where the
            # EOF should be. We need to at least write a single \0
            if ( $dstFh && $status eq 'sparse' ) {
                sysseek( $dstFh, $srcSeek - 1, SEEK_SET )
                  || $self->logger->logcroak('Cannot seek in dst file');
                syswrite( $dstFh, "\0" );
            }
        }

        if (   $self->truncate
            && $dstFh
            && $srcSeek
            && @{ $self->data } == 1
            && $self->data->[0]->{start} == 0
            && $self->data->[0]->{end} == 0 )
        {
            $self->logger->debug("Truncating dst file at <$srcSeek>");
            truncate( $dstFh, $srcSeek );

            my $chkSeek = ceil( $srcSeek / $self->bs ) * $hashSize;
            $self->logger->debug("Truncating chk file at <$chkSeek>");
            truncate( $chkFh, $chkSeek );
        }

    }
    catch {
        croak $_;
    }
    finally {
        # If we opened the files (we got string with path), then we close them
        # if we got a filehandle then we do nothing
        if ($srcClose) {
            $self->logger->debug( sprintf 'closing src file <%s>', $self->src );
        }
        if ($dstClose) {
            $self->logger->debug( sprintf 'closing dst file <%s>', $self->dst );
        }
        if ($chkClose) {
            $self->logger->debug( sprintf 'closing chk file <%s>', $self->chk );
        }
    };

    # Make Perl::Critic happy
    return;
}

################################################################

=begin comment

Private
Get file handle

=end comment

=cut

sub _getFh {
    my ( $self, $name, $file, $closeFile, $mode ) = @_;

    if ( my $t = reftype($file) ) {
        if ( $t eq 'GLOB' ) {
            $self->logger->debug(
                sprintf '%s is a file handle, using that directly', $name );
            return $file;
        }
        else {
            $self->logger->logcroak(
                sprintf '<%s> is not a supported type for %s',
                $t, $name );
        }
    }
    else {
        $self->logger->debug( sprintf 'opening %s file <%s>', $name, $file );
        sysopen( my $fh, $file, $mode )
          || $self->logger->logcroak( sprintf 'error opening <%s>', $file );
        ${$closeFile} = 1;
        return $fh;
    }

    # Make Perl::Critic happy
    croak 'We should never end here!';
}

################################################################

=head1 LICENSE AND COPYRIGHT

This software is copyright (c) 2019 by Thor Dreier-Hansen.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

Terms of the Perl programming language system itself:

=over

=item * the 
        L<GNU General Public License|http://dev.perl.org/licenses/gpl1.html>
        as published by the Free Software Foundation; either
        L<version 1|http://dev.perl.org/licenses/gpl1.html>,
        or (at your option) any later version, or

=item * the L<"Artistic License"|http://dev.perl.org/licenses/artistic.html>

=back

See L<http://dev.perl.org/licenses/> for more information.

=cut

1;    # End of IO::BlockSync
